src/mpi/coll/algorithms/common/coll_tree_util.h

```diff
@@ -0,0 +1,162 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+/* Routine to calculate log_k of an integer */
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_ilog(int k, int number)
+{
+    int i = 1, p = k - 1;
+
+    for (; p - 1 < number; i++)
+        p *= k;
+
+    return i;
+}
+
+/* Routing to calculate base^exp for integers */
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_ipow(int base, int exp)
+{
+    int result = 1;
+
+    while (exp) {
+        if (exp & 1)
+            result *= base;
+
+        exp >>= 1;
+        base *= base;
+    }
+
+    return result;
+}
+
+/* get the number at 'digit'th location in base k representation of 'number' */
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_getdigit(int k, int number, int digit)
+{
+    return (number / MPIR_ALGO_ipow(k, digit)) % k;
+}
+
+/* set the number at 'digit'the location in base k representation of 'number' to newdigit */
+MPL_STATIC_INLINE_PREFIX int MPIR_ALGO_setdigit(int k, int number, int digit, int newdigit)
+{
+    int res = number;
+    int lshift = MPIR_ALGO_ipow(k, digit);
+    res -= MPIR_ALGO_getdigit(k, number, digit) * lshift;
+    res += newdigit * lshift;
+    return res;
+}
```

Comment:
Maybe these math functions should go in MPL?

Reply:
I am not sure if such functions are useful anywhere other than in collectives.  But they can certainly go in coll/include/coll_util.h

Reply:
Even if they're not useful outside of collectives, they're general utility functions, that's still where they should go.

Reply:
I have currently included them in coll_util.h. There is a commit later on that adds a mpl/math directory. I will move these there after that commit.